### PR TITLE
Add visualization tool

### PR DESCRIPTION
**Pull request by Hailong-am**
### Description
Visualization Tool, coming from https://github.com/opensearch-project/ml-commons/pull/1649
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Hailong-am** [comment]
please help to add backport 2.x label also

File: src/main/java/org/opensearch/agent/tools/VisualizationsTool.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    @Override
+    public String getDescription() {
+        return description;
+    }
+
+    @Override
+    public void setDescription(String description) {
+        this.description = description;
+    }
+
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/agent/tools/VisualizationsTool.java:137]: Most of the other tools are using the Lombok `@Getter` and `@Setter` annotation.  You'll need the getters for type and version I think, and maybe add annotation for description?  Look at the other tools for confirmation.

File: src/test/resources/org/opensearch/agent/tools/visualization.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          ],
+          "migrationVersion": {
+            "visualization": "7.10.0"
+          },
+          "updated_at": "2023-11-10T02:50:24.881Z"
+        }
+      }
+    ]
+  }
+}
```
Comment (dbwiddis) [inline src/test/resources/org/opensearch/agent/tools/visualization.json:58]: Text files should end with a line break.

File: src/test/resources/org/opensearch/agent/tools/visualization_not_found.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  "hits": {
+    "total": {
+      "value": 0,
+      "relation": "eq"
+    },
+    "max_score": null,
+    "hits": []
+  }
+}
```
Comment (dbwiddis) [inline src/test/resources/org/opensearch/agent/tools/visualization_not_found.json:18]: Add a line break here too :)

**Review by dbwiddis — COMMENTED** [review]
LGTM with some nits.

File: src/main/java/org/opensearch/agent/tools/VisualizationsTool.java (hunk 1 above)
Comment (Hailong-am) [reply src/main/java/org/opensearch/agent/tools/VisualizationsTool.java]: thanks for suggestion, use lombok annotation now.

File: src/test/resources/org/opensearch/agent/tools/visualization.json (hunk 2 above)
Comment (Hailong-am) [reply src/test/resources/org/opensearch/agent/tools/visualization.json]: fixed

File: src/test/resources/org/opensearch/agent/tools/visualization_not_found.json (hunk 3 above)
Comment (Hailong-am) [reply src/test/resources/org/opensearch/agent/tools/visualization_not_found.json]: fixed

File: src/main/java/org/opensearch/agent/tools/VisualizationsTool.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@ToolAnnotation(VisualizationsTool.TYPE)
+public class VisualizationsTool implements Tool {
+    public static final String NAME = "Find Visualizations";
```
Comment (zane-neo) [inline src/main/java/org/opensearch/agent/tools/VisualizationsTool.java:38]: It looks tool's name are following camel case naming convention, is there any specific reason to add extra blank here?

Comment (Hailong-am) [reply src/main/java/org/opensearch/agent/tools/VisualizationsTool.java]: extra space removed

**Review by zane-neo — APPROVED** [review]

File: src/main/java/org/opensearch/agent/tools/VisualizationsTool.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+@Log4j2
+@ToolAnnotation(VisualizationsTool.TYPE)
+public class VisualizationsTool implements Tool {
+    public static final String NAME = "FindVisualizations";
+    public static final String TYPE = "VisualizationTool";
+    public static final String VERSION = "v1.0";
+
+    public static final String SAVED_OBJECT_TYPE = "visualization";
+    private static final String DEFAULT_DESCRIPTION =
+        "Use this tool to find user created visualizations. This tool takes the visualization name as input and returns the first 3 matching visualizations";
```
Comment (dhrubo-os) [inline src/main/java/org/opensearch/agent/tools/VisualizationsTool.java:44]: Does `3` needs to be static? Can't we do like 3 by default but the input is dynamic?

**Comment by dhrubo-os** [comment]
Need to resolve the conflict

File: src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java [hunk 6]
```diff
@@ -82,7 +82,8 @@ public <T> void run(Map<String, String> parameters, ActionListener<T> listener)
             ? Boolean.parseBoolean(parameters.get("highCardinality"))
             : null;
         final Long lastUpdateTime = parameters.containsKey("lastUpdateTime") && StringUtils.isNumeric(parameters.get("lastUpdateTime"))
-        ? Long.parseLong(parameters.get("lastUpdateTime")) : null;
+            ? Long.parseLong(parameters.get("lastUpdateTime"))
```
Comment (dhrubo-os) [inline src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java:85]: add a test?

**Comment by Hailong-am** [comment]
> Need to resolve the conflict

resolved

File: src/main/java/org/opensearch/agent/tools/VisualizationsTool.java (hunk 5 above)
Comment (Hailong-am) [reply src/main/java/org/opensearch/agent/tools/VisualizationsTool.java]: good suggestion, update the number to dynamic setting.

File: src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java (hunk 6 above)
Comment (Hailong-am) [reply src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java]: this is a code style update to make spotless happy.

File: src/main/java/org/opensearch/agent/tools/VisualizationsTool.java (hunk 5 above)
Comment (dhrubo-os) [reply src/main/java/org/opensearch/agent/tools/VisualizationsTool.java]: May be update the description as well?

Comment (Hailong-am) [reply src/main/java/org/opensearch/agent/tools/VisualizationsTool.java]: the specific number in description has been removed. now it's `"Use this tool to find user created visualizations. This tool takes the visualization name as input and returns the matching visualizations"
`

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**